In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
path = 'C:/Users/wjssm/Desktop/kaggle/Elo_Merchant_Category_Recommendation/data/'

In [3]:
train = pd.read_csv(path + 'train.csv') #key : card_id
test = pd.read_csv(path + 'test.csv') #key : card_id
history = pd.read_csv(path + 'historical_transactions.csv') #key : card_id, merchant_id
new_history = pd.read_csv(path + 'new_merchant_transactions.csv') #key : card_id, merchant_id
merchant = pd.read_csv(path + 'merchants.csv') #key : merchant_id

## card_id 기준

#### card_id 기준으로 history가 new_history 모두 포함

In [4]:
# new_history가 history의 부분집합인가 - card_id 기준
#포함관계 history <- new_history
set(new_history.card_id.unique()).issubset(set(history.card_id.unique()))

True

In [36]:
print('new_history에서 history card_id 뺀 수 : ',len(set(new_history.card_id.unique()) - set(history.card_id.unique())))
print('history에서 new_history card_id 뺀 수 : ',len(set(history.card_id.unique()) - set(new_history.card_id.unique())))

new_history에서 history card_id 뺀 수 :  0
history에서 new_history card_id 뺀 수 :  35539


In [6]:
# train이 history의 부분집합인가
set(train.card_id.unique()).issubset(set(history.card_id.unique()))

True

In [7]:
# test가 history의 부분집합인가
set(test.card_id.unique()).issubset(set(history.card_id.unique()))

True

In [8]:
# train 과 test의 중복 확인 : 전체 unique 합친거 - 합집합 개수
len(list(train.card_id.unique()) + list(test.card_id.unique())) - len(set(list(train.card_id.unique()) + list(test.card_id.unique())))

0

In [15]:
# train과 new history 중복 확인
len(list(new_history.card_id.unique()) + list(train.card_id.unique())) - len(set(list(new_history.card_id.unique()) + list(train.card_id.unique())))

179986

In [16]:
# test와 new history 중복 확인
len(list(new_history.card_id.unique()) + list(test.card_id.unique())) - len(set(list(new_history.card_id.unique()) + list(test.card_id.unique())))

110015

#### train + test = history (card_id)

In [29]:
#history
len(set(list(train.card_id.unique()) + list(test.card_id.unique())) - set(history.card_id))

0

In [26]:
#history
len(set(history.card_id) - set(train.card_id) - set(test.card_id))

0

### 결론

* history의 고유 인덱스의 개수 = train의 고유 인덱스의 개수 + test의 고유 인덱스의 개수
* new_history의 고유 인덱스는 history의 고유 인덱스의 부분집합 (new_history가 history의 부분집합이라는 소리가 아닙니다 !!)
* new_history와 history 사이에 중복된 행은 없음
* train의 고유 인덱스는 총 201917개, new_history와 train의 중복 인덱스는 총 179986개
* test의 고유 인덱스는 총 123623개, new_history와 test의 중복인덱스는 총 110015개
* 즉, train의 89%, test의 89%가 new_history와 관계있음

## merchant_id

In [62]:
#history 차집합 new_hitory
len(set(history.merchant_id) - set(new_history.merchant_id))

108504

In [63]:
#new_history 차집합 history
len(set(new_history.merchant_id) - set(history.merchant_id))

8322

In [64]:
 # history 교집합 new history
len(set(history.merchant_id.unique()).intersection(set(new_history.merchant_id.unique())))

217808

In [65]:
# history 합집합 new history
len(set(list(history.merchant_id.unique()) + list(new_history.merchant_id.unique())))

334634

In [61]:
108504+8322+217808

334634

In [43]:
print('new_history merchant_id null 개수 : ',new_history.merchant_id.isna().sum())
print('history merchant_id null 개수 : ',history.merchant_id.isna().sum())

new_history merchant_id null 개수 :  26216
history merchant_id null 개수 :  138481


#### merchant = new_history + history (merchant_id)

In [51]:
len(set(merchant.merchant_id) - set(new_history.merchant_id.dropna()) - set(history.merchant_id.dropna()))

0

In [54]:
len(set(list(new_history.merchant_id.dropna()) + list(history.merchant_id.dropna())) - set(merchant.merchant_id))

0

### 결론

* history의 merchant_id의 고유 인덱스는 merchant의 merchant_id의 고유 인덱스의 부분 집합 (단, null 제거 시)
* new_history의 merchant_id의 고유 인덱스는 merchant의 merchant_id의 고유 인덱스의 부분 집합 (단, null 제거 시)

In [66]:
len(set(history.merchant_id.dropna()) - set(merchant.merchant_id))

0

In [67]:
len(set(new_history.merchant_id.dropna()) - set(merchant.merchant_id))

0